### DISCUSSION TITLE
Training on custom dataset

### DISCUSSION THREAD

**Discussion by hgaiser** [STVIR/pysot#410] [General]
First of all, thank you for this implementation. I was hoping to train on a custom dataset, but couldn't find a suggested approach for this in the README. What is the best practice for training on my own data? Is there an advised format to convert my data to? Can I find this definition somewhere?

I'm comfortable writing some custom code for parsing my dataset, I was just hoping to get some insight on what the best practices are for custom training processes.

**Comment by ZhiyuanChen** [comment]
> First of all, thank you for this implementation. I was hoping to train on a custom dataset, but couldn't find a suggested approach for this in the README. What is the best practice for training on my own data? Is there an advised format to convert my data to? Can I find this definition somewhere?

Have you checkouted the [TRAIN](https://github.com/STVIR/pysot/blob/master/TRAIN.md)?

**Comment by hgaiser** [comment]
@zyc-ai I have, but it only mentions training on existing datasets:

- VID
- YOUTUBEBB
- DET
- COCO

I would like to train on my own dataset. I looked into using one of the existing formats, like VID, but it doesn't feel like this is the intended way. There are some assumptions about directory structure and file naming conventions (every filename should be an integer for example).

**Comment by ZhiyuanChen** [comment]
Ahhh yes, it would be nice if there are instructions for custom dataset. 
For now, the easiest way could be write your own subdataset class and add it to dataset. Another easy way is to just reorg dataset like ybb or vid and than use provided directly.

**Reply by ssbagalkar** [reply]
Hello @zyc-ai I have a custom dataset which I organized as COCO and trained it using just COCO style format by changing the config to just have COCO instead of all 4 -VID, YOUTUBEBB, DET  and COCO. Do you see any problem with this approach? I have seen my overall training loss decrease significantly using a pretrained backbone of ResNet-50 model. But it seems like model is overfitting and I want a way to check the validation loss and other characteristics like success and precision mentioned [here](https://github.com/ssbagalkar/pysot-toolkit) Is there a way I can check my results on validation and testing set for custom dataset? 

I have trained on approximately 5k train mages(prior to augmentation) for approximately 20(default) epochs. Here is my total training loss. 

Can you please suggest steps I can take to improve the results as when I run my model on test video, it does not track the object very well, but on the training video it is far better, suggesting to me it has overfitted. Thanks a lot
![loss](https://user-images.githubusercontent.com/21297086/93718006-ebc01680-fb3e-11ea-9a51-1416c0b903f8.png)

**Reply by ZhiyuanChen** [reply]
It is theoretically correct. However, to the best of my knowledge, COCO is an image dataset, which is probably not an ideal choice for training a tracking algorithm. The easiest way to do is to add more training data, especially video data like YBB. It would requires further insights of the failure cases to know better how to improve.

**Reply by ssbagalkar** [reply]
Agreed, but I have a `custom dataset` which is just in COCO format. How can I get to see the validation loss? I don't think validation loss is written to the SummaryWriter. Is that correct?

**Reply by ZhiyuanChen** [reply]
I think it's correct, you are more than welcome to submit a MR about this
